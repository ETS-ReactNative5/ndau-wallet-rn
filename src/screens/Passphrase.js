import React, { Component } from 'react'
import {
  StyleSheet,
  View,
  ScrollView,
  TextInput,
  Alert,
  Image,
  TouchableOpacity,
  Text,
  StatusBar,
  Platform
} from 'react-native'
import CommonButton from '../components/CommonButton'
import cssStyles from '../css/styles'
import MultiSafeHelper from '../helpers/MultiSafeHelper'
import RNExitApp from 'react-native-exit-app'
import { SafeAreaView } from 'react-navigation'
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp
} from 'react-native-responsive-screen'
import UserData from '../model/UserData'
import AppConstants from '../AppConstants'
import {
  NEW_WALLET_SETUP_TYPE,
  RECOVERY_WALLET_SETUP_TYPE
} from '../components/SetupProgressBar'
import FlashNotification from '../components/FlashNotification'
import Padding from '../components/Padding'
import OrderAPI from '../api/OrderAPI'
import AsyncStorageHelper from '../model/AsyncStorageHelper'
import styleConstants from '../css/styleConstants'
import FontAwesome5Pro from 'react-native-vector-icons/FontAwesome5Pro'
import WaitingForBlockchainSpinner from '../components/WaitingForBlockchainSpinner'

const NDAU = require('img/ndau_multi_large_1024.png')

class Passphrase extends Component {
  constructor (props) {
    super(props)

    this.state = {
      password: '',
      showErrorText: false,
      loginAttempt: 1,
      spinner: false
    }

    this.maxLoginAttempts = 10
  }

  login = async () => {
    this.setState({ spinner: true }, async () => {
      try {
        let user = await MultiSafeHelper.getDefaultUser(this.state.password)
        let marketPrice = 0
        if (user) {
          console.log(
            `user in Passphrase found is ${JSON.stringify(user, null, 2)}`
          )

          // cache the password
          await AsyncStorageHelper.setApplicationPassword(this.state.password)

          try {
            await UserData.loadData(user)
            marketPrice = await OrderAPI.getMarketPrice()
          } catch (error) {
            FlashNotification.showError(error.message, false, false)
            this.setState({ spinner: false })
            return
          }

          FlashNotification.hideMessage()
          this.setState({ spinner: false }, () => {
            this.props.navigation.navigate('Dashboard', {
              user,
              encryptionPassword: this.state.password,
              marketPrice
            })
          })
        } else {
          this.showLoginError()
          this.setState({ spinner: false })
        }
      } catch (error) {
        console.log(error)
        this.showLoginError()
        this.setState({ spinner: false })
      }
    })
  }

  showExitApp () {
    Alert.alert(
      '',
      `You have hit the maximum amount of login attempts.`,
      [
        {
          text: 'Exit app',
          onPress: () => {
            RNExitApp.exitApp()
          }
        }
      ],
      { cancelable: false }
    )
  }

  showLoginError = () => {
    if (this.state.loginAttempt === this.maxLoginAttempts) {
      this.showExitApp()
    }
    FlashNotification.showError(
      `Login attempt ${this.state.loginAttempt} of ${
        this.maxLoginAttempts
      } failed.`
    )
    this.setState({ loginAttempt: this.state.loginAttempt + 1 })
  }

  showInformation = () => {
    Alert.alert(
      'Information',
      'Please enter the password you chose to encrypt this app. ' +
        'This is not the same thing as your six-character ID or key ' +
        'recovery phrase.',
      [{ text: 'OK', onPress: () => {} }],
      { cancelable: false }
    )
  }

  showSetup = async () => {
    FlashNotification.hideMessage()
    this.props.navigation.navigate('SetupWelcome', {
      walletSetupType: NEW_WALLET_SETUP_TYPE
    })
  }

  showPasswordReset = user => {
    FlashNotification.hideMessage()
    this.props.navigation.navigate('SetupGetRecoveryPhrase', {
      user: user,
      mode: AppConstants.PASSWORD_RESET_MODE,
      walletSetupType: RECOVERY_WALLET_SETUP_TYPE
    })
  }

  dropDownSelected = (index, value) => {
    this.setState({
      userId: value
    })
  }

  render () {
    console.log(`rendering Passphrase`)
    const { textInputColor } = this.state
    return (
      <SafeAreaView style={cssStyles.safeContainer}>
        <StatusBar barStyle='light-content' backgroundColor='#1c2227' />
        <View style={cssStyles.container}>
          <ScrollView style={cssStyles.contentContainer}>
            <WaitingForBlockchainSpinner spinner={this.state.spinner} />
            <Padding top={2}>
              <View style={styles.imageView}>
                <Image style={styles.image} source={NDAU} />
              </View>
            </Padding>

            <Padding top={2}>
              <View style={{ flexDirection: 'row' }}>
                <TextInput
                  style={{
                    height: hp('7%'),
                    width: wp('96%'),
                    borderColor: 'gray',
                    borderWidth: 1,
                    borderRadius: 3,
                    paddingLeft: wp('1%'),
                    color: '#000000',
                    backgroundColor: '#ffffff',
                    fontSize: 18,
                    fontFamily: 'TitilliumWeb-Regular'
                  }}
                  onChangeText={password => this.setState({ password })}
                  value={this.state.password}
                  placeholder='App Password'
                  placeholderTextColor='#333'
                  secureTextEntry={!this.state.showPasswords}
                  autoCapitalize='none'
                />
              </View>
            </Padding>

            <Padding top={0}>
              <Padding>
                <View style={styles.centerTextView}>
                  <Text
                    onPress={this.showPasswordReset}
                    style={cssStyles.linkText}
                  >
                    Forgot your password?
                  </Text>
                </View>
              </Padding>
              <Padding>
                <View style={styles.imageView}>
                  <TouchableOpacity onPress={this.showInformation}>
                    <Image
                      style={{ width: 35, height: 38 }}
                      source={require('img/info_icon_gold.png')}
                    />
                  </TouchableOpacity>
                </View>
              </Padding>

              {this.state.showErrorText ? (
                <Padding>
                  <View style={styles.errorContainer}>
                    <Text style={cssStyles.errorText}>
                      Please enter the passphrase you chose to decrypt this app.{' '}
                    </Text>
                  </View>
                </Padding>
              ) : null}
            </Padding>
          </ScrollView>
          <View style={cssStyles.footer}>
            <CommonButton onPress={this.login} title='Login' />
          </View>
        </View>
      </SafeAreaView>
    )
  }
}

const styles = StyleSheet.create({
  button: {},
  textContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  imageView: {
    justifyContent: 'center',
    alignItems: 'center'
  },
  centerTextView: {
    justifyContent: 'center',
    alignItems: 'center'
  },
  image: {
    width: wp('100%'),
    ...Platform.select({
      ios: {
        height: hp('20%')
      },
      android: {
        height: hp('30%')
      }
    })
  },
  infoIcon: {
    marginLeft: 12
  }
})

export default Passphrase
