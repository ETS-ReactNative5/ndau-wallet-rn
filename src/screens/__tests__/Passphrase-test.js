import { mount } from 'enzyme'
import React from 'react'
import { StyleSheet } from 'react-native'
import Passphrase from '../Passphrase'

import renderer from 'react-test-renderer'

describe('testing Passphrase...', () => {
  let styles = StyleSheet.create({
    wizardText: {
      color: '#ffffff',
      fontSize: 20
    }
  })
  const navigator = {
    setStyle: () => {},
    toggleNavBar: () => {}
  }

  it('can click the button', () => {
    const wrapper = mount(
      <Passphrase navigator={navigator} parentStyles={styles} />
    )
    const onlyButton = wrapper.find('Button').at(0)
    onlyButton.simulate('click')
  })
})
