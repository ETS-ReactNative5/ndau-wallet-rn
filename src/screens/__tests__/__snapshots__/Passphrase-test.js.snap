// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Passphrase... renders correctly 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "backgroundColor": "#1c2227",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#1c2227",
        "flex": 1,
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      }
    }
  >
    <RCTScrollView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "paddingBottom": 40,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../img/n_icon_ko.png",
              }
            }
            style={
              Object {
                "tintColor": "#4e957a",
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "justifyContent": "flex-end",
              "margin": 10,
            }
          }
        >
          <View
            onLayout={[Function]}
          >
            <View
              accessible={true}
              hitSlop={
                Object {
                  "bottom": 6,
                  "left": 4,
                  "right": 4,
                  "top": 6,
                }
              }
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              pointerEvents="box-only"
            >
              <View
                onResponderRelease={[Function]}
                onStartShouldSetResponder={[Function]}
                pointerEvents="none"
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "borderBottomColor": "rgba(255, 255, 255, 1)",
                      "borderBottomWidth": 0.5,
                      "height": 66,
                      "paddingBottom": 8,
                      "paddingTop": 31,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "position": "absolute",
                        "top": 35.5,
                      }
                    }
                  >
                    <Text
                      accessible={true}
                      allowFontScaling={true}
                      ellipsizeMode="tail"
                      numberOfLines={1}
                      style={
                        Object {
                          "color": "rgba(255, 255, 255, 1)",
                          "fontSize": 18,
                        }
                      }
                    >
                      Please choose a User ID
                    </Text>
                  </View>
                  <View
                    style={
                      Object {
                        "flexDirection": "row",
                      }
                    }
                  >
                    <TextInput
                      absoluteRTLLayout={false}
                      allowFontScaling={true}
                      autoCapitalize="sentences"
                      data={Array []}
                      disableFullscreenUI={true}
                      dropdownMargins={
                        Object {
                          "max": 16,
                          "min": 8,
                        }
                      }
                      dropdownOffset={
                        Object {
                          "left": 0,
                          "top": 32,
                        }
                      }
                      editable={false}
                      itemColor="rgba(0, 0, 0, .54)"
                      itemCount={4}
                      itemPadding={8}
                      itemTextStyle={
                        Object {
                          "color": "#ffffff",
                          "fontFamily": "TitilliumWeb-Regular",
                          "fontSize": 18,
                        }
                      }
                      labelExtractor={[Function]}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onChangeText={[Function]}
                      onContentSizeChange={[Function]}
                      onFocus={[Function]}
                      propsExtractor={[Function]}
                      renderAccessory={[Function]}
                      rippleDuration={400}
                      selectedItemColor="#000000"
                      selectionColor="rgb(0, 145, 234)"
                      shadeOpacity={0.12}
                      style={
                        Array [
                          Object {
                            "flex": 1,
                            "margin": 0,
                            "padding": 0,
                            "top": 2,
                          },
                          Object {
                            "color": "#ffffff",
                            "fontSize": 18,
                            "height": 27,
                            "textAlign": "left",
                          },
                          undefined,
                        ]
                      }
                      underlineColorAndroid="transparent"
                      useNativeDriver={false}
                      value=""
                      valueExtractor={[Function]}
                    />
                    <View
                      style={
                        Object {
                          "alignSelf": "flex-start",
                          "justifyContent": "center",
                          "top": 2,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "alignItems": "center",
                            "height": 24,
                            "justifyContent": "center",
                            "width": 24,
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "backgroundColor": "transparent",
                              "height": 6,
                              "overflow": "hidden",
                              "width": 12,
                            }
                          }
                        >
                          <View
                            style={
                              Array [
                                Object {
                                  "height": 8,
                                  "transform": Array [
                                    Object {
                                      "translateY": -4,
                                    },
                                    Object {
                                      "rotate": "45deg",
                                    },
                                  ],
                                  "width": 8,
                                },
                                Object {
                                  "backgroundColor": "#ffffff",
                                },
                              ]
                            }
                          />
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
                <View
                  style={
                    Object {
                      "flexDirection": "row",
                      "height": 8,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "flex-start",
                          "bottom": 0,
                          "left": 0,
                          "paddingVertical": 4,
                          "position": "absolute",
                          "right": 0,
                          "top": 0,
                        }
                      }
                    >
                      <Text
                        accessible={true}
                        allowFontScaling={true}
                        ellipsizeMode="tail"
                        numberOfLines={1}
                        style={
                          Object {
                            "backgroundColor": "transparent",
                            "color": "rgb(213, 0, 0)",
                            "fontSize": 0,
                            "opacity": 0,
                          }
                        }
                      />
                    </View>
                    <View
                      style={
                        Object {
                          "alignItems": "flex-start",
                          "bottom": 0,
                          "left": 0,
                          "paddingVertical": 4,
                          "position": "absolute",
                          "right": 0,
                          "top": 0,
                        }
                      }
                    >
                      <Text
                        accessible={true}
                        allowFontScaling={true}
                        ellipsizeMode="tail"
                        numberOfLines={1}
                        style={
                          Object {
                            "backgroundColor": "transparent",
                            "color": "#ffffff",
                            "fontSize": 12,
                            "opacity": 1,
                          }
                        }
                      />
                    </View>
                  </View>
                </View>
              </View>
              <View
                accessible={true}
                onLayout={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onRippleAnimation={[Function]}
                onStartShouldSetResponder={[Function]}
                pointerEvents="box-only"
                rippleCentered={false}
                rippleColor="#ffffff"
                rippleDuration={400}
                rippleFades={true}
                rippleOpacity={0.54}
                rippleSequential={true}
                rippleSize={0}
                style={
                  Object {
                    "height": 51,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 16,
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "bottom": 0,
                        "left": 0,
                        "overflow": "hidden",
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      },
                      Object {
                        "borderRadius": 0,
                      },
                    ]
                  }
                />
              </View>
            </View>
            <Modal
              hardwareAccelerated={false}
              onRequestClose={[Function]}
              supportedOrientations={
                Array [
                  "portrait",
                  "portrait-upside-down",
                  "landscape",
                  "landscape-left",
                  "landscape-right",
                ]
              }
              transparent={true}
              visible={false}
            >
              <View
                onResponderRelease={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "opacity": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <View
                  onStartShouldSetResponder={[Function]}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "rgba(255, 255, 255, 1.0)",
                        "borderRadius": 2,
                        "position": "absolute",
                        "shadowColor": "rgba(0, 0, 0, 1.0)",
                        "shadowOffset": Object {
                          "height": 2,
                          "width": 0,
                        },
                        "shadowOpacity": 0.54,
                        "shadowRadius": 2,
                      },
                      Object {
                        "height": 16,
                        "left": undefined,
                        "top": undefined,
                        "transform": Array [
                          Object {
                            "translateY": -51,
                          },
                        ],
                        "width": undefined,
                      },
                      undefined,
                    ]
                  }
                >
                  <RCTScrollView
                    contentContainerStyle={
                      Object {
                        "paddingVertical": 8,
                      }
                    }
                    data={Array []}
                    disableVirtualization={false}
                    getItem={[Function]}
                    getItemCount={[Function]}
                    horizontal={false}
                    initialNumToRender={10}
                    keyExtractor={[Function]}
                    maxToRenderPerBatch={10}
                    numColumns={1}
                    onContentSizeChange={[Function]}
                    onEndReachedThreshold={2}
                    onLayout={[Function]}
                    onMomentumScrollEnd={[Function]}
                    onScroll={[Function]}
                    onScrollBeginDrag={[Function]}
                    onScrollEndDrag={[Function]}
                    renderItem={[Function]}
                    scrollEnabled={false}
                    scrollEventThrottle={50}
                    stickyHeaderIndices={Array []}
                    style={
                      Object {
                        "borderRadius": 2,
                        "flex": 1,
                      }
                    }
                    updateCellsBatchingPeriod={50}
                    viewabilityConfigCallbackPairs={Array []}
                    windowSize={21}
                  >
                    <View />
                  </RCTScrollView>
                </View>
              </View>
            </Modal>
          </View>
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
              "marginLeft": 10,
              "marginRight": 10,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Enter your app password"
            placeholderTextColor="#333"
            secureTextEntry={true}
            style={
              Object {
                "backgroundColor": "#ffffff",
                "borderColor": "gray",
                "borderWidth": 1,
                "color": undefined,
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 18,
                "height": 45,
                "marginBottom": 10,
                "marginTop": 10,
                "paddingLeft": 10,
                "width": "93%",
              }
            }
            value=""
          />
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontSize": 20,
                  },
                  Object {
                    "marginLeft": 12,
                    "marginTop": 20,
                  },
                  Object {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "justifyContent": "flex-end",
          "margin": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginBottom": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          onPress={[Function]}
          style={
            Object {
              "color": "#dea85a",
              "fontFamily": "TitilliumWeb-Regular",
              "fontSize": 18,
              "textDecorationLine": "underline",
            }
          }
        >
          Create a new user
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginBottom": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#ffffff",
              "fontFamily": "TitilliumWeb-Regular",
              "fontSize": 18,
            }
          }
        >
          or
        </Text>
      </View>
      <View
        accessible={true}
        focusAnimation={null}
        focusAnimationDuration={225}
        hitSlop={
          Object {
            "bottom": 6,
            "left": 4,
            "right": 4,
            "top": 6,
          }
        }
        onLayout={[Function]}
        onPress={[Function]}
        onPressIn={[Function]}
        onPressOut={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        pointerEvents="box-only"
        rippleCentered={false}
        rippleColor="#ffffff"
        rippleDuration={400}
        rippleFades={true}
        rippleOpacity={0.3}
        rippleSequential={true}
        rippleSize={0}
        style={
          Object {
            "backgroundColor": "#4e957a",
            "borderRadius": 2,
            "height": 36,
            "justifyContent": "space-around",
            "minWidth": 64,
            "paddingHorizontal": 8,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "fontSize": 14,
                "fontWeight": "500",
                "textAlign": "center",
              },
              Object {
                "color": "#ffffff",
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 14,
              },
              Object {
                "color": "#ffffff",
              },
            ]
          }
        >
          LOGIN
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#ffffff",
              "borderRadius": 2,
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
        <View
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "bottom": 0,
                "left": 0,
                "overflow": "hidden",
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              Object {
                "borderRadius": 2,
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
